pkg/kubelet/kubelet.go

```diff
@@ -1739,8 +1749,8 @@ func (kl *Kubelet) syncPod(pod *api.Pod, mirrorPod *api.Pod, podStatus *kubecont
 	}
 	kl.statusManager.SetPodStatus(pod, apiPodStatus)
 
-	// Kill pods we can't run.
-	if err := canRunPod(pod); err != nil || pod.DeletionTimestamp != nil {
+	// Kill pods we can't run, pods pending deletion, or pods that have been marked as failed.
```

Comment:
Note that there is a race condition -- once the status is set to "failed", it's possible that the HandlePodCleanup could pick up this pod and attempt to kill it. We need to add checks in that HandlePodCleanup, so it doesn't attempt to kill the pod if a pod worker exists. Of course, if we have a internal pod object with states and access methods, that'd solve the problem too (which @vishh wants).